core/src/main/scala/zio/jdbc/ZConnection.scala

```diff
@@ -27,18 +27,22 @@ import java.sql.{ Connection, PreparedStatement, Statement }
  */
 final class ZConnection(private[jdbc] val underlying: ZConnection.Restorable) extends AnyVal {
 
-  def access[A](f: Connection => A): ZIO[Any, Throwable, A] =
-    ZIO.attemptBlocking(f(underlying))
+  def access[A](f: Connection => A): IO[FailedAccess[A], A] =
+    ZIO.scoped {
+      accessZIO(f.andThen(ZIO.attempt(_)))
+    }
 
-  def accessZIO[A](f: Connection => ZIO[Scope, Throwable, A]): ZIO[Scope, Throwable, A] =
-    ZIO.blocking(f(underlying))
+  def accessZIO[A](f: Connection => ZIO[Scope, Throwable, A]): ZIO[Scope, FailedAccess[A], A] =
+    ZIO.blocking(f(underlying)).refineOrDie { case e: SQLException =>
+      FailedAccess(e, f)
```

Comment:
Question: What the idea behind wrapping with `FailedAccess`? As a user the way pool works is an implementation detail, do I want to see an underlying SQL error directly? Is my assumption correct, that the pool is expected not to fail due to its own error, only in result of some underlying error?

Reply:
I think that `access` and `accessZIO` are designed to be used also by the user, so a `FailedAccess` would return the error and the function that the user has employed. But when used internally it might be better to handle the error before getting to the user.

Reply:
I see you point of `accessZIO` being public API. Still not sure what's the benefit of wrapping in a separate case class `FailedAccess`. Assuming error handling logic is implemented up in the stack, along with `FailedToConnection`, now the user have to pattern match over multiple case classes in order to extract error details: `FailedAccess`, `FailedToConnect`, and `ZSQLException`. Plus, the only place `accessZIO` is used so far is [`executeSqlWith`](https://github.com/zio/zio-jdbc/pull/142/files#diff-e55904c7e5721f68784499255320debb11655107d1afdc52e5d65216941dd58eR80) which extracts `SQLException` and "dies" on other failures. 

I think `ZSQLException` makes more sense, but in this case we need to figure out how to handle user code errors in `f`. Can we limit its error channel to `SQLException` or `ZSQLException` (by requiring to `refineOrDie` in `f`). My problem with this approach is that if we consider `accessZIO` as publish API (i.e. `f` is use code), we should allow to use type error. No sure extractly why, but for some reason I still like `Throwable` at this level (maybe because the wrapper make the code of `executeSqlWith` more complicated and causes more issues in error handling code). 🤔

Reply:
I also think that ideally `accessZIO` should return only `ZSQLException`. At this point, `FailedAccess` exists merely because `accessZIO` is public.

In reality, using `accessZIO` in unconventional ways that do not return only `SQLException` should not be standard. We could refine `accessZIO` and create an `unsafeAccessZIO` that returns Throwable. And this way we can avoid `FailedAccess`. What do you think?

Reply:
> I also think that ideally accessZIO should return only ZSQLException. At this point, FailedAccess exists merely because accessZIO is public and the user might want to use it in other ways.

I think `accessZIO`, is a low level API to access `Connection`, so `Throwable` is ok, and `access` cannot even express user specific errors as it can only use `ZIO.attempt`. I think we can consider anything up to `ZResultSet` as semi-private API of the lib. 🤔

Reply:
I agree with @domartynov